Preserve config/routes.rb and config/locales/en.yml files when `rails app:update`

ryohashimoto opened the pull request:
### Summary

When `rails app:update` is executed, config/routes.rb and config/locales/en.yml files would be overwritten by almost empty template files. This behavior is not what we expected in Rails upgrade process because they were edited so many times by framework users like files in app directory. They have many application logic and custom messages not like other configuration files.

I think rails app:update will be more usable if these files will not be overwritten because in many software, user generated contents are not erased during upgrading process.

I modified the `config_when_updating` method in `Rails::AppBuilder` and these config files are no longer target files for the `app:update` task.

ryohashimoto committed:
Preserve config/routes.rb and config/locales/en.yml files when `rails app:update`

Suppressing overwrites with almost empty file contents.

y-yagi commented:
Certainly not many, but `config/locales/en.yml` is updated occasionally. d4ddf7b72f52597d0c0f90794cf52f6d0d687d4e, 957e3d496683c8e7862a2acf8637def0b483bac9. 
I include these changes at update time. So I am against the exclusion of `config/locales/en.yml`. 

`config/routes.rb` certainly does not think that there is a change that needs to be taken for now. 
But that's the same for other configs. It seems confusing to treat only `config/route.rb` specially. 

Also, if do not use the `config` method, we will need to care `config_when_updating` as well when adding config files. It is not maintainable(and will surely forget to add it).

kaspth commented:
One thing that could be interesting is to diff the routes and locale files from the last release and stashing that in a yml file in the packaged gem when `rake release` is run. Something like:

```yml
# railties/changed_configurations.yml # Or somewhere else.
- config/routes.rb
- config/locales/en.yml
```

ryohashimoto commented:
The reason why do not want to overwrite these files is because they were created by framework users like files in  app directory. They have many business logic and custom messages and should not be overwritten like other configuration files.

I think `rails app:update` will be more usable if these files will not be overwritten because in many software, user generated contents are not erased during upgrading process.

@y-yagi 

> Certainly not many, but config/locales/en.yml is updated occasionally

Yes, the comments of the file have being updated, but the actual content of the file has not been upgraded since old versions.

> if do not use the config method, we will need to care config_when_updating

It's difficult to preserve files using `config` method. So I copied and customized for the purpose.

kaspth commented:
> The reason why do not want to overwrite these files is because they were edited so many times by framework users like files in app directory. They have many application logic and custom messages not like other configuration files.
> 
> I think `rails app:update` will be more usable if these files will not be overwritten because in many software, user generated contents are not erased during upgrading process.

Yes, we get the reason already, thank you. We're just saying that the proposed solution has maintainability problems for us, so it won't be acceptable as is.

y-yagi commented:
It should be left up to the user to decide if a comment is necessary, and I think that it is not good to decide on the framework side. 
Anyway, I agree that this approach has maintainability concerned.

By the way, I agree that overwrite is unuseful. 
So I added `merge` action to `thor` and uses this when files conflicted in `app:update`.
https://github.com/erikhuda/thor/pull/605

Is this not helpful in your case?

ryohashimoto commented:
Thanks for your comments and suggestions. I will fix maintainability problems later.

ryohashimoto committed:
use options[:update] in Rails::AppBuilder#config

- if option[:update] is set, config/routes.rb and config/locales files will
  not be overwritten.
- In order to correctly set option[:update] to true, set { update: true } in test cases.

ryohashimoto commented:
To solve the maintainability issue and to eliminate the duplication of some codes,  I added the commit 4712254f01e30779fef8c43239e5e077030f0a6f .
In this commit,  `options[:update]` is used for checking if the `config` method is in upgrading process or not.

@kaspth

Could you check the commit? 

@y-yagi 

Thanks, I checked your commit in Thor repository.
To merge conflicted files, user can choose `"m"` option from prompt with `[Ynaqdhm]`.
It is useful for advanced user, but still in a difficult situation for end-user who doesn't know the option.
Because the user could not see new contents before choosing the option and by default (`"Y"` option) the files will be overwritten.

y-yagi commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -114,7 +114,7 @@ def config
         directory "environments"
         directory "initializers"
-        directory "locales"
+        directory "locales" unless options[:update]
```
Since this needs to be fixed manually when `locals` changes, there are maintainability problems still. 

We need to be able to do this check dynamically. 
(I also think that the approach of calculating differences with `rake release` is appropriate)

ryohashimoto replied:
`directory "locales"` just copy the directory existing in [template folder](https://github.com/rails/rails/tree/master/railties/lib/rails/generators/rails/app/templates).
Like environments and initializers directories specified before the line.
Please tell me a little more about the maintainability problems you mentioned.

ryohashimoto replied:
If more locale files will be added in future rails versions, we will need to implement some codes in `config_when_updating` method like recently added files under the `config/initializers` directory. I think it's not a problem.

y-yagi commented:
> Because the user could not see new contents before choosing the option and by default ("Y" option) the files will be overwritten.

Thanks for your feedback! 

Does this mean that you want to see the changes before doing the merge process, then how about adding an option (like show) for that? 
Also, if it is not good to overwrite the default, how about making it changeable?

ryohashimoto commented:
@y-yagi

> Thor has diff action, is that not enough?
> Also, if it is not good to overwrite the default, how about making it changeable?

Not enough. I proposed the PR because `config/routes.rb` and `config/locales` files needs to be treated as files in `app` directory in upgrading process.
